[Docs]: Add docs for `schema` in block API

dmsnell opened the pull request:
## Description

While the `schema` property is mentioned in the block API documentation
for `raw` transforms it's not explained.

In this patch we're adding a short sub-section to expand on the purpose
and implementation of the `schema` in a way that can hopefully give
developers a stronger starting point for when they need to implement
new raw transformations.

## How has this been tested?

This is a documentation-only change and does not touch the code.
Please review the new explanations and provide feedback on the
wording and example and whether or not you find the structure of
inlining the subsection into the existing block API document to be
a helpful move, or whether you think it belongs elsewhere.

## Types of changes
This change adds an explanation to existing block API documentation.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

sarayourfriend commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -373,6 +373,52 @@ transforms: {
+<h4 id="schemas-and-content-models">Schemas and Content Models</h4>
+
+When pasting content it's possible to define a [content model](https://html.spec.whatwg.org/multipage/dom.html#content-models) that will be used to validate and process pasted content.
+It's often the case that HTML pasted into the editor will contain a mixutre of elements that _should_ transfer as well as elements that _shouldn't_.
```
```suggestion
It's often the case that HTML pasted into the editor will contain a mixture of elements that _should_ transfer as well as elements that _shouldn't_.
```

dmsnell replied:
thanks for the catch!

dmsnell replied:
> I'm not familiar at all with how this works technically. From my view this is a helpful explanation of this concept, but I can't verify that it is technically correct or isn't missing something critical 😅

That's the point of this PR 😄

I've attempted to explain what the code actually does. It could be wrong, but I have tried to be accurate here and only explain things I confirmed.

sarayourfriend replied:
Makes sense! I'm hoping someone with more familiarity can confirm there aren't any nuances that should be included. From a "know nothing" perspective like my own this is a huge improvement over the non-existent previous documentation 😁

sarayourfriend reviewed (commented):
I'm not familiar at all with how this works technically. From my view this is a helpful explanation of this concept, but I can't verify that it is technically correct or isn't missing something critical 😅

dmsnell committed:
[Docs]: Add docs for schema in block API

While the `schema` property is mentioned in the block API documentation
for `raw` transforms it's not explained.

In this patch we're adding a short sub-section to expand on the purpose
and implementation of the `schema` in a way that can hopefully give
developers a stronger starting point for when they need to implement
new raw transformations.

Co-authored-by: sarayourfriend <24264157+sarayourfriend@users.noreply.github.com>

dmsnell commented:
Rebased against the latest `trunk` - is there anyone opposing this update?

dmsnell commented:
Flakey test suites :sigh:

dmsnell commented:
@ellatrix do you have any thoughts on this doc update?

bph reviewed (approved):
This is a great addition to the page section!  It will educate developers well about how transforms can be applied. Thanks. 
I don't have any edits to suggest as I am not deep enough into the development parts (yet) but I understood the gist of it, and remember enough to take a look here when the need arises.

gwwar commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -221,6 +221,52 @@ transforms: {
+When pasting content it's possible to define a [content model](https://html.spec.whatwg.org/multipage/dom.html#content-models) that will be used to validate and process pasted content.
+It's often the case that HTML pasted into the editor will contain a mixture of elements that _should_ transfer as well as elements that _shouldn't_.
+For example, consider pasting `<span class="time">12:04 pm</span>` into the editor.
+We want to copy `12:04 pm` and omit the `<span>` and its `class` attribute because those won't carry the same meaning or structure as they originally did from where they were copied.
```
Should we add an example for the 12:04pm scenario, for what the raw paste handler is doing?

dmsnell replied:
yeah that's a good idea and it'll show the non-function form of the property. I'll add it in here tersely and update. thanks for calling that out.

gwwar commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -221,6 +221,52 @@ transforms: {
+We want to tell the editor to allow the inner `h2` and `p` elements.
+We do this by supplying the following schema.
+In this example we're using the function form, which accepts an argument supplying `phrasingContentSchema` (as well as a boolean `isPaste` indicating if the transformation operation started with pasting text).
+The `phrasingContentSchema` is pre-defined to match HTML phrasing elements, such as `<strong>` and `<sup>` and `<kbd>`.
```
My first instinct was to look up what `phrasingContentSchema` actual shape was. This is defined in https://github.com/WordPress/gutenberg/blob/trunk/packages/dom/src/phrasing-content.js

dmsnell replied:
when I wrote this up I debated adding the actual shape but was worried it wouldn't be maintained, which is why I ended up referring to `@wordpress/dom` for the description of the schema.

do you have thoughts on this? if it weren't something that was specifically using the HTML terminology and concepts I would lean more on describing it here, but I wanted to stress the links to the HTML specs over our code in order to pass people to better and more robust documentation.

gwwar replied:
I don't think we need to provide the actual shape, but I'm thinking most folks might not understand what this does without a few examples.

gwwar reviewed (approved):
Good first start here @dmsnell thanks for adding documentation! ✨  I left some minor notes but nothing blocking.

I suspect a lot of folks probably learn from examples, so I wonder if we could setup a sort of interactive transform/paste sandbox in the future.

dmsnell commented:
> I wonder if we could setup a sort of interactive transform/paste sandbox in the future.

this one is going to be out of scope for me right now. the trigger for this work was me needing a `schema` but having no idea why or how, so I created notes as I figured it out and wanted to share them.

apart from that I'll add the 12:04pm example and hopefully we can get this in for future improvements.

dmsnell committed:
Review feedback: add example

dmsnell committed:
Reference full URls for source code since this will be in the handbook.

gwwar commented:
> this one is going to be out of scope for me right now.

No worries on that. I was mostly thinking of what'd be nice to have for future us. For example whenever I'm looking at a new area like understanding transform options, I tend to lookup actual instances and play around in a development branch.
